crucible-llvm/src/Lang/Crucible/LLVM/MemModel/Generic.hs

```diff
@@ -130,52 +133,92 @@ data MemWrite sym
 tgAddPtrC :: (1 <= w, IsExprBuilder sym) => sym -> NatRepr w -> LLVMPtr sym w -> Addr -> IO (LLVMPtr sym w)
 tgAddPtrC sym w x y = ptrAdd sym w x =<< constOffset sym w y
 
+-- | An environment used to interpret 'OffsetExpr's, 'IntExpr's, and 'Cond's.
+-- These data structures may contain uninterpreted variables to be filled in
+-- with the offset address of a load or store, or the size of the current
+-- region. Since regions may be unbounded in size, the size argument is a
+-- 'Maybe' type.
+data ExprEnv sym w = ExprEnv { loadOffset  :: SymBV sym w
+                             , storeOffset :: SymBV sym w
+                             , sizeData    :: Maybe (SymBV sym w) }
+
+-- | Interpret an 'OffsetExpr' as a 'SymBV'. Although 'OffsetExpr's may contain
+-- 'IntExpr's, which may be undefined if they refer to the size of an unbounded
+-- memory region, this function will throw an error if both (1) the 'sizeData'
+-- in the 'ExprEnv' is 'Nothing' and (2) 'StoreSize' occurs anywhere in the
+-- 'OffsetExpr'.
 genOffsetExpr ::
   (1 <= w, IsSymInterface sym) =>
   sym -> NatRepr w ->
-  (SymBV sym w, SymBV sym w, SymBV sym w) ->
-  OffsetExpr -> IO (SymBV sym w)
-genOffsetExpr sym w f@(load, store, _size) expr =
+  ExprEnv sym w ->
+  OffsetExpr ->
+  IO (SymBV sym w)
+genOffsetExpr sym w f@(ExprEnv load store _size) expr =
   case expr of
     OffsetAdd pe ie -> do
       pe' <- genOffsetExpr sym w f pe
       ie' <- genIntExpr sym w f ie
-      bvAdd sym pe' ie'
+      case ie' of
+        Nothing -> error "Cannot construct an offset that references the size of an unbounded region"
```

Comment:
We should use `panic` here instead of `error` if we believe that this is a programming bug/error for this situation to occur.  If, instead, we believe that a user input could elicit this situation, we need some more robust error handling.

Reply:
I think it should be a bug for this to occur; there are specific times that `StoreSize` is referenced in an `OffsetBV`, for example in `Crucible.LLVM.MemModel.Common.symbolicRangeLoad`, but this function is only invoked (in `MemModel.Generic`) when the size is bounded. Because these `OffsetExpr`s are constructed in a separate place from where they are interpreted, however, this property is hard to verify.

I will change this to a `panic`.